### PR TITLE
 many: use changes + tasks for quota group operations

### PR DESCRIPTION
**Pull request by anonymouse64**
The exported methods from the servicestate package all return tasksets now, 
which is expected to be put into changes that are executed by the overlord 
loop. This involves changes in many parts that use quotas, such as the tests,
where a new mock function to create a quota group in state without running any
servicectl commands as well as changes to the client side.

### CODE REVIEW

File: daemon/api_quotas.go [hunk 1]
```diff
@@ -169,27 +174,37 @@ func postQuotaGroup(c *Command, r *http.Request, _ *auth.UserState) Response {
...
 			return BadRequest(err.Error())
 		}
+		chgSummary = "Remove quota group"
 	default:
 		return BadRequest("unknown quota action %q", data.Action)
 	}
-	return SyncResponse(nil)
+
+	chg := newChange(st, "quota-control", chgSummary, []*state.TaskSet{ts}, data.Snaps)
+	st.EnsureBefore(0)
```
Comment (pedronis) [inline daemon/api_quotas.go:208]: we tend to use a mockable helper ensureStateSoon

File: daemon/api_quotas_test.go [hunk 2]
```diff
@@ -59,14 +59,17 @@ func (s *apiQuotaSuite) SetUpTest(c *check.C) {
 
 	// POST requires root
 	s.expectedWriteAccess = daemon.RootAccess{}
+
+	d.Overlord().Loop()
```
Comment (pedronis) [inline daemon/api_quotas_test.go:63]: that might avoid to do this and also allow to check that EnsureBefore was called

File: overlord/servicestate/quota_control.go [hunk 3]
```diff
@@ -81,21 +81,72 @@ func quotaGroupsAvailable(st *state.State) error {
 	return nil
 }
 
+// MockQuotaInState mocks a given quota group in the state without ensuring any
+// of the on-disk state exists. It is meant to be consumed only by tests.
+func MockQuotaInState(st *state.State, name string, parentName string, snaps []string, memoryLimit quantity.Size) error {
```
Comment (pedronis) [inline overlord/servicestate/quota_control.go:86]: I'm really not a fan of this living here, we might have to introduce both an internal and a servicestatetest package to avoid this though.  devicestate does that for similar reasons

File: overlord/servicestate/quota_handlers.go [hunk 4]
```diff
@@ -322,22 +325,26 @@ type ensureSnapServicesForGroupOptions struct {
...
 
-	if meter == nil {
-		meter = progress.Null
+	var meterLocked, meterUnlocked progress.Meter
+	if t == nil {
+		meterLocked = progress.Null
+		meterUnlocked = progress.Null
+	} else {
+		meterUnlocked = snapstate.NewTaskProgressAdapterUnlocked(t)
+		meterLocked = snapstate.NewTaskProgressAdapterUnlocked(t)
```
Comment (pedronis) [inline overlord/servicestate/quota_handlers.go:341]: this should be s/Unlocked/Locked/. It's probably a bit annoying to check for this in tests:/

**Review by pedronis — COMMENTED** [review]
did a pass, found a theoretical issue but I don't think it relates to the spread tests issues, as afaict systemd uses the meters for real only for Stop/Restart

File: overlord/servicestate/quota_handlers.go (hunk 4 above)
Comment (anonymouse64) [reply overlord/servicestate/quota_handlers.go:341]: this error actually comes from https://github.com/snapcore/snapd/pull/10410, but you're right it's wrong... I will adjust that there. like you said though, not sure how to really test the different meters

File: overlord/servicestate/quota_control.go (hunk 3 above)
Comment (anonymouse64) [reply overlord/servicestate/quota_control.go:86]: hmm, I'm not sure how to move this function to another package without exporting things we probably don't want to export from this package or without duplicating the implementation, for example this function uses `quotaCreateImpl` which uses `patchQuotas`, so either I could export one or both of those functions or I could duplicate the implementation of one or both of those functions.

I guess the other thing I could do is some sort of hook situation, but that gets weird to do since servicestatetest needs to import servicestate, so any hooks that are exported in servicestatetest can't be assigned to from servicestate.

Comment (pedronis) [reply overlord/servicestate/quota_control.go:86]: that's why I mentioned that you would probably need to introduce a servicestate/internal package. I can try to explore this a bit more concretely

Comment (pedronis) [reply overlord/servicestate/quota_control.go:86]: I did the prerequisites steps of what I described/had in mind in https://github.com/snapcore/snapd/pull/10429

File: daemon/api_quotas.go (hunk 1 above)
Comment (anonymouse64) [reply daemon/api_quotas.go]: done

File: daemon/api_quotas_test.go (hunk 2 above)
Comment (anonymouse64) [reply daemon/api_quotas_test.go]: done

File: overlord/servicestate/quota_control.go (hunk 3 above)
Comment (anonymouse64) [reply overlord/servicestate/quota_control.go]: this helper is now in servicestatetest, but I introduced it as part of a smaller PR: https://github.com/snapcore/snapd/pull/10434 so please review that one first

**Comment by anonymouse64** [comment]
Unfortunately this is blocked now, after reviewing the spread test failures again, it seems that changes like those in https://github.com/snapcore/snapd/pull/10420/commits/763dd017ba4bdc126f307544e03a8eaa4714bdb3 are papering over a real issue.

The issue is that we now have a task which gets marked done before it is done doing things. It is okay in the case where we get rebooted in the middle after unlocking the state and come back and re-execute the task, it will not break anything, but we now have race conditions in the following situation (which is what the spread test is doing):

1. `quota-control` task is executing
1. `snap set-quota ...` command is looping waiting for the change with the `quota-control` task in it to be in the `Done` state
1. quota-control task finishes running `wrappers.EnsureSnapServices()` and marks itself done, and releases the state lock
1. the client now sees that the change is `Done`, so it returns success
1. The client now assumes that the group is created/updated/removed and that all affected services have been updated, however they have not been and future ops executed immediately fail, such as checking that a service was restarted or killed (due to OOM with the low memory limit like in the test) or etc 
1. The quota-control task continues executing, since there is more code after `wrappers.EnsureSnapServices()` that needs to be executed, mainly that we restart the services with the state unlocked since restarting snap services is a snap-controlled time option (meaning it could take an arbitrary amount of time due to `start-timeout` and `stop-timeout`, etc.)
1. If the client now tries to check / execute the things in 5) above, they pass since now the task is fully executed


There are a few ways we could solve this, but this is a bit of an unfortunate corner case. Options I can think of right now:

1. Accept that life is hard and computers triply so and just adjust the spread test to check the things we want to check in a loop
2. Introduce a new task type and execute the `systemctl restart ...` bits that we need to run with state unlocked in the second task, this ensures that the first task runs `wrappers.EnsureSnapServices()` and when it marks itself Done, it will not be run again, leading to problems. Then the change will need to wait for the whole thing to finish, but if the second task gets interrupted or executed again after a reboot it's not a big deal since the operations are idempotent anyways.
3. Adjust tasks / changes in some other way so that we can mark the task as Done() but not mark the entire change as Done() until all the handler functions have actually returned, but this is a behavior change and may have other unintended effects across the codebase and also would need to be split out into it's own PR separate from this

**Comment by pedronis** [comment]
This is now rebased on #10458 which should solve the issues we had with the spread tests. At this point the two last commits cancel each other, I have both to keep history but could also be dropped.

**Comment by pedronis** [comment]
Conflict detection is implemented in #10471 based on this one

File: overlord/servicestate/quota_control.go [hunk 5]
```diff
@@ -85,18 +85,41 @@ func quotaGroupsAvailable(st *state.State) error {
...
+	}
+
+	// make sure the group does not exist yet
+	if _, ok := allGrps[name]; ok {
+		return nil, fmt.Errorf("group %q already exists", name)
+	}
+
+	if memoryLimit == 0 {
+		return nil, fmt.Errorf("cannot create quota group with no memory limit set")
+	}
```
Comment (pedronis) [inline overlord/servicestate/quota_control.go:107]: these new precond checks are not tested

File: overlord/servicestate/quota_control.go [hunk 6]
```diff
@@ -85,18 +85,41 @@ func quotaGroupsAvailable(st *state.State) error {
...
+
+	// make sure the memory limit is at least 4K, that is the minimum size
+	// to allow nesting, otherwise groups with less than 4K will trigger the
+	// oom killer to be invoked when a new group is added as a sub-group to the
+	// larger group.
+	if memoryLimit <= 4*quantity.SizeKiB {
+		return nil, fmt.Errorf("memory limit for group %q is too small: size must be larger than 4KB", name)
```
Comment (pedronis) [inline overlord/servicestate/quota_control.go:114]: same

File: overlord/servicestate/quota_control.go [hunk 7]
```diff
@@ -85,18 +85,41 @@ func quotaGroupsAvailable(st *state.State) error {
...
 	}
 
-	// TODO: switch to returning a taskset with the right handler instead of
-	// executing this directly
+	// make sure the specified snaps exist and aren't currently in another group
+	if err := validateSnapForAddingToGroup(st, snaps, name, allGrps); err != nil {
+		return nil, err
```
Comment (pedronis) [inline overlord/servicestate/quota_control.go:119]: same

File: overlord/servicestate/quota_control.go [hunk 8]
```diff
@@ -105,46 +128,57 @@ func CreateQuota(st *state.State, name string, parentName string, snaps []string
...
+	// make sure the group exists
+	grp, ok := allGrps[name]
+	if !ok {
+		return nil, fmt.Errorf("cannot remove non-existent quota group %q", name)
+	}
+
+	// XXX: remove this limitation eventually
+	if len(grp.SubGroups) != 0 {
+		return nil, fmt.Errorf("cannot remove quota group with sub-groups, remove the sub-groups first")
 	}
```
Comment (pedronis) [inline overlord/servicestate/quota_control.go:167]: same

File: overlord/servicestate/quota_control.go [hunk 9]
```diff
@@ -164,33 +198,56 @@ type QuotaGroupUpdate struct {
...
+		// full details
+		if updateOpts.NewMemoryLimit < grp.MemoryLimit {
+			return nil, fmt.Errorf("cannot decrease memory limit of existing quota-group, remove and re-create it to decrease the limit")
+		}
+	}
+
+	// now ensure that all of the snaps mentioned in AddSnaps exist as snaps and
+	// that they aren't already in an existing quota group
+	if err := validateSnapForAddingToGroup(st, updateOpts.AddSnaps, name, allGrps); err != nil {
+		return nil, err
```
Comment (pedronis) [inline overlord/servicestate/quota_control.go:232]: same

**Review by pedronis — COMMENTED** [review]
lgtm but the new preconditions checks are not tested afaict

File: client/quota.go [hunk 10]
```diff
@@ -95,12 +97,9 @@ func (client *Client) RemoveQuotaGroup(groupName string) error {
...
 	var body bytes.Buffer
 	if err := json.NewEncoder(&body).Encode(data); err != nil {
-		return err
-	}
-	if _, err := client.doSync("POST", "/v2/quotas", nil, nil, &body, nil); err != nil {
-		return err
+		return "", err
 	}
-	return nil
+	return client.doAsync("POST", "/v2/quotas", nil, nil, &body)
```
Comment (stolowski) [inline client/quota.go:102]: Nitpick, EnsureQuota above decorates the error from client.doAsync with extra context, should we have "cannot remove quota group: %w" here for consistency and better error report?

File: overlord/servicestate/quota_control.go [hunk 11]
```diff
@@ -164,33 +198,56 @@ type QuotaGroupUpdate struct {
...
-	// TODO: switch to returning a taskset with the right handler instead of
-	// executing this directly
+	// check that the memory limit is not being decreased
+	if updateOpts.NewMemoryLimit != 0 {
+		// we disallow decreasing the memory limit because it is difficult to do
+		// so correctly with the current state of our code in
+		// EnsureSnapServices, see comment in ensureSnapServicesForGroup for
+		// full details
+		if updateOpts.NewMemoryLimit < grp.MemoryLimit {
+			return nil, fmt.Errorf("cannot decrease memory limit of existing quota-group, remove and re-create it to decrease the limit")
```
Comment (stolowski) [inline overlord/servicestate/quota_control.go:225]: AFAICT this case isn't covered either.

**Review by stolowski — APPROVED** [review]
Looks good. One nitpick and a few test cases missing as pointed out by Samuele (and I think I found one more), but I'm fine if they are addressed in a followup not to block this PR (and also because it's already big).

File: overlord/servicestate/quota_control.go (hunk 11 above)
Comment (pedronis) [reply overlord/servicestate/quota_control.go:225]: I think I now pushed tests that cover most of these

**Review by pedronis — COMMENTED** [review]
+1

**Review by stolowski — APPROVED** [review]
The last commit adding extra tests looks good, thank you!

File: client/quota.go (hunk 10 above)
Comment (anonymouse64) [reply client/quota.go:102]: seems a reasonable suggestion for a followup to me

**Review by anonymouse64 — COMMENTED** [review]
I agree with Pawel's suggestion, otherwise looks good to me, thanks for the additional tests and pre-req PR's that have already landed

**Review by pedronis — APPROVED** [review]
actually vote

File: client/quota.go (hunk 10 above)
Comment (anonymouse64) [reply client/quota.go:102]: done in https://github.com/snapcore/snapd/pull/10484
